### ISSUE TITLE
Interlocks caused by transaction management

### ISSUE THREAD

**Issue by jersiovic** [OrchardCMS/OrchardCore#8705]
### Describe the bug
Since we started to use orchard one year ago for a software as a service project, I'm facing interlock problems that I barely patch adding extra commits. Specialyl if we make load test. However after a while I see more interlock problems in the logs. 
This problem was more obvious when we added integration tests that execute in parallel vs an orchard instance. Each of those tests use different users and subsets of data, so in theory despite they use same server one test don't have side effects on the other.
After one or two runs of the set tests some random tests fail because of interlock exceptions.

Taking a look on how Orchard Core handles transactions and sessions I see that as it happened in Orchard 1, Orchard Core always provides us with a session which always open a transaction if a query to the database is done with that session.
And before the session is disposed if no one committed changes, changes are committed. 

IMO it is a handy mechanism for devs that can forget about transaction in almost all the cases. But the disadvantage,is it increases the probability of interlock and puts more pressure on db cause it has to manage more locks on tables than what it is actually needed.

In general all the code related with  stores is candidate to produce interlock in a multithread environment if it is used enough concurrently, I mean: UserStore, OpenIdApplicationStore, OpenITokenStore, OpenIdAuthorizationStore, ... . Why? Cause no matter for what purpose they are used their read operations always open a transaction. And the classes which uses it are third party code that are not aware of transaction thing, because the interfaces they use for using them doesn't offer a way for choosing queries with transaction or without transaction depending on the scenario they are implementing.
In my case I got crazy trying to understand why locks exactly happen because if each test works with a subset of different records I didn't see why it should happen. Finally I arrived to the believe that maybe sql server was blocking morre records per table than those affected by the queries cause for performance reasons it ends locking blocks of records when many locks happens at the same time on the same table. 

### To Reproduce

It is hard to me due time constrainst to provide a sample showing the problem based only on Orchard source. But I'm going to provide a PR to Orchard Core and another to YesSQL that solved my issues in my project, just in case you think it is useful.

### Expected behavior
It should not produce interlock exception

### Screenshots
exception logged
```
2021-03-01 09:21:41.1870|[21]|Default||7485e2b0-4b522b054abad907.||App.UsersWebAPI.EventHandlers.ProviderUserChangesEventHandlerOnUser|ERROR|Error añadiendo usuarios provider en usuarios Microsoft.Data.SqlClient.SqlException (0x80131904): Transaction (process ID 52) was deadlocked on lock resources with another process and has been chosen as the deadlock victim. Rerun the transaction.
   at Microsoft.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at Microsoft.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at Microsoft.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at Microsoft.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at Microsoft.Data.SqlClient.SqlCommand.FinishExecuteReader(SqlDataReader ds, RunBehavior runBehavior, String resetOptionsString, Boolean isInternal, Boolean forDescribeParameterEncryption, Boolean shouldCacheForAlwaysEncrypted)
   at Microsoft.Data.SqlClient.SqlCommand.CompleteAsyncExecuteReader(Boolean isInternal, Boolean forDescribeParameterEncryption)
   at Microsoft.Data.SqlClient.SqlCommand.InternalEndExecuteNonQuery(IAsyncResult asyncResult, Boolean isInternal, String endMethod)
   at Microsoft.Data.SqlClient.SqlCommand.EndExecuteNonQueryInternal(IAsyncResult asyncResult)
   at Microsoft.Data.SqlClient.SqlCommand.EndExecuteNonQueryAsync(IAsyncResult asyncResult)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
--- End of stack trace from previous location where exception was thrown ---
   at Dapper.SqlMapper.ExecuteMultiImplAsync(IDbConnection cnn, CommandDefinition command, IEnumerable multiExec) in /_/Dapper/SqlMapper.Async.cs:line 654
   at YesSql.Session.FlushAsync() in D:\PROYECTOS\AppServer\yessql\src\YesSql.Core\Session.cs:line 711
   at YesSql.Services.DefaultQuery.Query`1.FirstOrDefaultImpl() in D:\PROYECTOS\AppServer\yessql\src\YesSql.Core\Services\DefaultQuery.cs:line 912
   at OrchardCore.Users.Services.UserStore.FindByNameAsync(String normalizedUserName, CancellationToken cancellationToken) in D:\PROYECTOS\AppServer\Orchard2\src\OrchardCore\OrchardCore.Users.Core\Services\UserStore.cs:line 121
   at Microsoft.AspNetCore.Identity.UserManager`1.FindByNameAsync(String userName)
   at Microsoft.AspNetCore.Identity.UserValidator`1.ValidateUserName(UserManager`1 manager, TUser user, ICollection`1 errors)
   at Microsoft.AspNetCore.Identity.UserValidator`1.ValidateAsync(UserManager`1 manager, TUser user)
   at Microsoft.AspNetCore.Identity.UserManager`1.ValidateUserAsync(TUser user)
   at Microsoft.AspNetCore.Identity.UserManager`1.UpdateUserAsync(TUser user)
   at App.UsersWebAPI.EventHandlers.ProviderUserChangesEventHandlerOnUser.AddedUsersToAProviderAsync(AddedUsersToAProviderContext context, Action`2 addError) in D:\PROYECTOS\AppServer\App.Modules\App.UsersWebAPI\1 EventHandlers\ProviderUserChangesEventHandlerOnUser.cs:line 59
ClientConnectionId:041acec6-ee6d-4af3-a4e1-d3a5342ff069
Error Number:1205,State:51,Class:13    at Microsoft.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at Microsoft.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at Microsoft.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at Microsoft.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at Microsoft.Data.SqlClient.SqlCommand.FinishExecuteReader(SqlDataReader ds, RunBehavior runBehavior, String resetOptionsString, Boolean isInternal, Boolean forDescribeParameterEncryption, Boolean shouldCacheForAlwaysEncrypted)
   at Microsoft.Data.SqlClient.SqlCommand.CompleteAsyncExecuteReader(Boolean isInternal, Boolean forDescribeParameterEncryption)
   at Microsoft.Data.SqlClient.SqlCommand.InternalEndExecuteNonQuery(IAsyncResult asyncResult, Boolean isInternal, String endMethod)
   at Microsoft.Data.SqlClient.SqlCommand.EndExecuteNonQueryInternal(IAsyncResult asyncResult)
   at Microsoft.Data.SqlClient.SqlCommand.EndExecuteNonQueryAsync(IAsyncResult asyncResult)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
--- End of stack trace from previous location where exception was thrown ---
   at Dapper.SqlMapper.ExecuteMultiImplAsync(IDbConnection cnn, CommandDefinition command, IEnumerable multiExec) in /_/Dapper/SqlMapper.Async.cs:line 654
   at YesSql.Session.FlushAsync() in D:\PROYECTOS\AppServer\yessql\src\YesSql.Core\Session.cs:line 711
   at YesSql.Services.DefaultQuery.Query`1.FirstOrDefaultImpl() in D:\PROYECTOS\AppServer\yessql\src\YesSql.Core\Services\DefaultQuery.cs:line 912
   at OrchardCore.Users.Services.UserStore.FindByNameAsync(String normalizedUserName, CancellationToken cancellationToken) in D:\PROYECTOS\AppServer\Orchard2\src\OrchardCore\OrchardCore.Users.Core\Services\UserStore.cs:line 121
   at Microsoft.AspNetCore.Identity.UserManager`1.FindByNameAsync(String userName)
   at Microsoft.AspNetCore.Identity.UserValidator`1.ValidateUserName(UserManager`1 manager, TUser user, ICollection`1 errors)
   at Microsoft.AspNetCore.Identity.UserValidator`1.ValidateAsync(UserManager`1 manager, TUser user)
   at Microsoft.AspNetCore.Identity.UserManager`1.ValidateUserAsync(TUser user)
   at Microsoft.AspNetCore.Identity.UserManager`1.UpdateUserAsync(TUser user)
   at App.UsersWebAPI.EventHandlers.ProviderUserChangesEventHandlerOnUser.AddedUsersToAProviderAsync(AddedUsersToAProviderContext context, Action`2 addError) in D:\PROYECTOS\AppServer\App.Modules\App.UsersWebAPI\1 EventHandlers\ProviderUserChangesEventHandlerOnUser.cs:line 59
```

**Comment by deanmarcussen** [comment]
This is an interesting one, and useful that you are able to reproduce with your integration tests.

An initial thought.

We recently updated the indexes on a lot of these tables, to include the `DocumentId` property, which seems to have resolved some deadlock related issues, as it reduced table scanning. See https://github.com/OrchardCMS/OrchardCore/issues/8518

I know you've probably tweaked a lot of your indexes, have you been including the `DocumentId` on them?

> Finally I arrived to the believe that maybe sql server was blocking morre records per table than those affected by the queries cause for performance reasons it ends locking blocks of records when many locks happens at the same time on the same table.

The hint from google around this says this is exactly right, and also suggests that better non clustered indexes are a way to resolve it. (but maybe you already have done this)

**Comment by jersiovic** [comment]
OMG haha we forgot to add that field to indexes, 
I don't now if this will solve the problem but independently ot it, it is clearly is a great improvement.
Thank you guys!

I will try and will keep you informed of the results.
Independently of that the conversation in the PR I sent I think is still worth to have it https://github.com/sebastienros/yessql/pull/349

**Comment by deanmarcussen** [comment]
Ha cool!

For infos, we discovered that without a `DocumentId` in the indexes, it would always do a table scan of the index, when doing a `Query<Model,Index>` because of the `Join` 

But when added the `DocumentId` it would be a non-clustered scan, rather than the full table. So it may help with deadlocks.

Will it solve everything? I don't know. But you've got the failing tests to check it, so lets see.

Look forward to hearing more info back.

**Comment by jtkech** [comment]
I cross my fingers 🤞

**Comment by jersiovic** [comment]
It worked, even removing all my changes for supporting readonly sessions. Only with changes needed for adding DocumentId on all the indexes the problem is fixed. No more interlock exceptions on integration tests!! 

My only doubt now is if it is interesting to support in yessql readonly queries (I mean with no transaction for scenaros of heavy load where even indexrd doesn't fix things)

**Comment by jtkech** [comment]
👍

**Comment by deanmarcussen** [comment]
Thats great @jersiovic 

Well worth that big pr then @jtkech good work!

I'll leave it with you a Seb on YesSql to continue the readonly feature

**Comment by deanmarcussen** [comment]
Where are we at with your other pr for the indexes?

Did what @jtkech do satisfy all those requirements?

**Comment by jersiovic** [comment]
> Where are we at with your other pr for the indexes?
> 
> Did what @jtkech do satisfy all those requirements?
I don't get exactly what you mean.

My orchard source code is very old in comparison with you dev branch, so instead of merging your changes I did them myself cause there have been many changes in indexes with my base branch.
I added DocumentId field to all indexes defined for MapIndext at  src/OrchardCore.Modules/OrchardCore.Users/Migrations.cs
and /src/OrchardCore/OrchardCore.OpenId.Core/YesSql/Migrations/OpenIdMigrations.cs
And for MapReduce indexes I also added sth like this cause those index tables have a different structure. I mention this cause I didn't see those indexes at @jtkech commits:
```
SchemaBuilder.AlterTable($"{nameof(OpenIdAppByRoleNameIndex)}_Document", table =>
            {
                table.CreateIndex($"IX_{nameof(OpenIdAppByRoleNameIndex)}_Document_{nameof(OpenIdAppByRoleNameIndex.RoleName)}", "DocumentId", $"{nameof(OpenIdAppByRoleNameIndex)}Id");
            });
```
I didn't do all the other changes made by @jtkech in other modules, cause we don't use those modules for out app.

**Comment by sebastienros** [comment]
I agree with @jersiovic that the index should also be added on the bridge tables (for reduce indexes).

**Comment by sebastienros** [comment]
@jersiovic Can you add a unit test in yessql that would show the issue?

Also it seems that FKs don't create indexes as I would have thought. So I'll need to add these indexes on DocumentId (since they are not the primary keys).

For the bridge table, we need to check if yessql is doing two joins or a join with AND. (Document.Id -> B.DocumentId + B.Id -> Index.Id). And change it if necessary, and also create the index that matches the query.

**Comment by jersiovic** [comment]
I tried months ago to reproduce problem in a plain simple test for showing you the problem but I was not able. I will try another time but using UserManager and a UserStore based on YesSql that seems to be what makes the difference.
Maybe if we isolate that scenario, next taking a look to source code o UserManager we can isolate the problem without it

**Comment by deanmarcussen** [comment]
Maybe the difference in some of the tests here https://github.com/sebastienros/yessql/blob/dd587b433c990dc18973957794aad88d868fd5f3/test/YesSql.Tests/CoreTests.cs#L3798-L3846

is that they are all query based, whereas with your issue, it is because of session commits, and table updates, rather than just queries

**Comment by jersiovic** [comment]
I have been trying for a while to reproduce interlock problems adding new unit tests to YesSql source which even used an entity User and also the UserManager, to mimic parts of the code causing problems in our app. However I didn't managed to reproduce the problems in unit tests executed in paralallel. Problems only happen on more complex scenarios on our app.

Recently I've updated orchard sources that use new YesSql version that doesn't open a transaction just for read, and things have changed for better. Now we don't suffer interlock exceptions thats's good.
However now tests fail randomly due to other exceptions that seems to be related to inconsistent data. Those exceptions happen related to our own custom documents, but also related with Orchard Users. Just for giving you more info take into account that we execute integration tests in parallel towards one server and each test creates at least one unique user.
This a sample of exception:
```
2021-06-04 13:12:42.6640|[22]|Default|00-a8609c445dfa4a47888caf48ac71b819-318db8c29fd6fb46-00||OrchardCore.Users.Services.UserStore|ERROR|Unexpected error while updating a user. Microsoft.Data.SqlClient.SqlException (0x80131904): The INSERT statement conflicted with the FOREIGN KEY constraint "FK_UserByRoleNameIndex_Document_Id". The conflict occurred in database "ktpTests", table "dbo.UserByRoleNameIndex", column 'Id'.
   at Microsoft.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at Microsoft.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at Microsoft.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at Microsoft.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at Microsoft.Data.SqlClient.SqlDataReader.TryHasMoreRows(Boolean& moreRows)
   at Microsoft.Data.SqlClient.SqlDataReader.TryHasMoreResults(Boolean& moreResults)
   at Microsoft.Data.SqlClient.SqlDataReader.TryNextResult(Boolean& more)
   at Microsoft.Data.SqlClient.SqlDataReader.NextResult()
   at YesSql.Commands.CreateIndexCommand.<AddToBatch>b__6_0(DbDataReader dr)
   at YesSql.Commands.BatchCommand.ExecuteAsync(DbConnection connection, DbTransaction transaction, ISqlDialect dialect, ILogger logger)
   at YesSql.Session.FlushAsync()
   at YesSql.Session.FlushAsync()
   at YesSql.Session.SaveChangesAsync()
   at YesSql.Session.SaveChangesAsync()
   at OrchardCore.Users.Services.UserStore.UpdateAsync(IUser user, CancellationToken cancellationToken)
ClientConnectionId:726153ff-4424-4c91-aa1d-5e3710b59d40
Error Number:547,State:0,Class:16    at Microsoft.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at Microsoft.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at Microsoft.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at Microsoft.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at Microsoft.Data.SqlClient.SqlDataReader.TryHasMoreRows(Boolean& moreRows)
   at Microsoft.Data.SqlClient.SqlDataReader.TryHasMoreResults(Boolean& moreResults)
   at Microsoft.Data.SqlClient.SqlDataReader.TryNextResult(Boolean& more)
   at Microsoft.Data.SqlClient.SqlDataReader.NextResult()
   at YesSql.Commands.CreateIndexCommand.<AddToBatch>b__6_0(DbDataReader dr)
   at YesSql.Commands.BatchCommand.ExecuteAsync(DbConnection connection, DbTransaction transaction, ISqlDialect dialect, ILogger logger)
   at YesSql.Session.FlushAsync()
   at YesSql.Session.FlushAsync()
   at YesSql.Session.SaveChangesAsync()
   at YesSql.Session.SaveChangesAsync()
   at OrchardCore.Users.Services.UserStore.UpdateAsync(IUser user, CancellationToken cancellationToken)
```

The solution has been easy but maybe it is not the better, I've changed default isolation level from read uncommitted to read committed. Just with this change the test seems to works properly by the moment. What do you think of that?

With previous version of YesSql to use read committed was not acceptable cause every read operation to the db would have been involved in a transaction and this would have caused more interlocks, don't you think?
But with the new version that only uses transactions for write operations it is acceptable to use read committed.

However I would prefer to be able to make it work with read uncommitted to reduce pressure over db server.
I understand that the right thing is to harness the `checkConcurrency` parameter provided by `Session.Save()` method. However I use it specifically for those documents for which I want to be sure all concurrent changes are applied to the document one after the other with a retry strategy for each operation if YesSql.ConcurrencyException exception is thrown. 
So, it sounds too much to have to use it for other documents that I don't mind if two parallel request change at he same time and the last tha does the operation wins. As it happens with User.

Any idea @sebastienros ?

**Comment by jersiovic** [comment]
> It worked, even removing all my changes for supporting readonly sessions. Only with changes needed for adding DocumentId on all the indexes the problem is fixed. No more interlock exceptions on integration tests!! 😄 😄
> 
> My only doubt now is if it is interesting to support in yessql readonly queries (I mean with no transaction for scenarios of heavy load where even indexes doesn't fix things)

**Comment by jersiovic** [comment]
> It worked, even removing all my changes for supporting readonly sessions. Only with changes needed for adding DocumentId on all the indexes the problem is fixed. No more interlock exceptions on integration tests!! 😄 😄
> 
Point that despite I said that on 2 Mar, later after adding more test interlock exceptions appeared randomly but with less frecuency and only on some machines, that was an extra reason for me for  spending time looking for fix the interlock problem without success till now, ater upgrade orchard source and change default isolation level

**Comment by jersiovic** [comment]
This is crarzy, after changing default isolation level to Read Uncommitted tests has started to run properly, and on some retries they fail and in others no. So I can say finally that isolation level doesn't seem to change the scenario.

**Comment by jersiovic** [comment]
I close this issue cause it seems interlock problems dissapeared after changes in transactions managemente done by @sebastienros at yessql :)
Current problem I was describing is related to another issue I've described here: https://github.com/OrchardCMS/OrchardCore/issues/9639
